test/views/components/RenderFormSpec.scala

```diff
@@ -0,0 +1,164 @@
+import org.scalatestplus.play.guice.GuiceOneAppPerSuite
+import play.api.data.Forms.nonEmptyText
+import play.api.i18n.{Lang, Messages, MessagesApi}
+import play.api.inject.Injector
+import play.api.mvc.AnyContentAsEmpty
+import play.api.test.FakeRequest
+import play.twirl.api.Html
+import views.html.components
+
+import scala.collection.JavaConverters._
+
+class RenderFormSpec extends ViewSpec with ViewFns with GuiceOneAppPerSuite {
+  private trait Test {
+    implicit val labels: Labels = LabelCache()
+    private def injector: Injector = app.injector
+    def messagesApi: MessagesApi = injector.instanceOf[MessagesApi]
+//    def formProvider: SubmittedAnswerFormProvider = injector.instanceOf[SubmittedAnswerFormProvider]
+    val formProvider = new SubmittedTextAnswerFormProvider()
+    implicit val fakeRequest: FakeRequest[AnyContentAsEmpty.type] = FakeRequest("GET", "/")
+    implicit def messages: Messages = messagesApi.preferred(fakeRequest)
+  }
+
+  private trait WelshTest extends Test {implicit override def messages: Messages = messagesApi.preferred(Seq(Lang("cy")))}
+
+  "English render_form" must {
+
+    "render a text input for a CurrencyInput" in new Test {
+      val currencyInput: CurrencyInput = models.ui.CurrencyInput(Text("Bank balance?", "Welsh, Bank balance?"), None, Seq.empty)
+      val page: FormPage = models.ui.FormPage("/url", currencyInput)
+      implicit val ctx: PageContext = models.PageContext(page, None, "sessionId", None, Text(), "processId", "processCode", labels)
+      val html: Html = components.render_form(currencyInput, "currency", formProvider("test" -> nonEmptyText))
+      val input: Element = getSingleElementByTag(html, "Input")
+      input.id() shouldBe "currency-0"
+    }
+
+    "render a text input for a CurrencyInputPoundsOnly" in new Test {
+      val currencyInput: CurrencyPoundsOnlyInput = models.ui.CurrencyPoundsOnlyInput(Text("Bank balance?", "Welsh, Bank balance?"), None, Seq.empty)
+      val page: FormPage = models.ui.FormPage("/url", currencyInput)
+      implicit val ctx: PageContext = models.PageContext(page, None, "sessionId", None, Text(), "processId", "processCode", labels)
+      val html: Html = components.render_form(currencyInput, "currency", formProvider("test" -> nonEmptyText))
+      val h1: Element = getSingleElementByTag(html, "H1")
+      h1.text() shouldBe "Bank balance?"
+      val input: Element = getSingleElementByTag(html, "Input")
+      input.id() shouldBe "currency-0"
+    }
+
+    "render three text inputs for a DateInput" in new Test {
+      val dateInput: DateInput = models.ui.DateInput(Text("Date of birth?", "Welsh, Date of birth?"), None, Seq.empty)
+      val page: FormPage = models.ui.FormPage("/url", dateInput)
+      implicit val ctx: PageContext = models.PageContext(page, None, "sessionId", None, Text(), "processId", "processCode", labels)
+      val html: Html = components.render_form(dateInput, "dateOfBirth", formProvider("test" -> nonEmptyText))
```

Comment:
In the case of date input the form provider is actually SubmittedDateAnswerFormProvider. So I would use

val formProvider = new SubmittedDateAnswerFormProvider()
val form = formProvider()

In the code this is done programmatically, but in the tests you have to create the form provider manually.

Reply:
updated